Kitodo-DataManagement/src/main/java/org/kitodo/data/database/beans/Keyworder.java

```diff
@@ -0,0 +1,512 @@
+        } catch (IOException | RuntimeException e) {
+            logger.catching(e instanceof FileNotFoundException ? Level.INFO : Level.WARN, e);
+            return emptyResult;
+        }
+    }
+
+    /**
+     * Returns a map for ruleset key translations. A cache is used, but if there
+     * is nothing in cache, the ruleset is parsed and the map is created. Since
+     * Kitodo-DataEditor is not available here, we have to do this directly, and
+     * it also increases performance massively.
+     * 
+     * @param file
+     *            indicates a ruleset
+     * @return a map
+     */
+    private static Map<String, Collection<String>> getRulesetLabelMap(String file) {
+        Map<String, Collection<String>> rulesetLabelMap = rulesetCache.get(file);
+        if (Objects.nonNull(rulesetLabelMap)) {
+            return rulesetLabelMap;
+        }
+        try {
+            File rulesetFile = Paths.get(KitodoConfig.getParameter("directory.rulesets"), file).toFile();
+            logger.debug("Reading {} ...", rulesetFile);
+            String ruleset = FileUtils.readFileToString(rulesetFile, StandardCharsets.UTF_8);
+            rulesetLabelMap = new HashMap<>();
+            Matcher keysMatcher = RULESET_KEY_PATTERN.matcher(ruleset);
+            Set<String> labels = new HashSet<>();
+            while (keysMatcher.find()) {
+                String key = normalize(keysMatcher.group(1));
+                Matcher labelMatcher = RULESET_LABEL_PATTERN.matcher(keysMatcher.group(2));
+                while (labelMatcher.find()) {
+                    labels.add(normalize(labelMatcher.group(1)));
+                }
+                rulesetLabelMap.put(key, labels);
+            }
+            rulesetCache.put(file, rulesetLabelMap);
+            return rulesetLabelMap;
+        } catch (IOException | RuntimeException e) {
+            logger.catching(Level.WARN, e);
+            return Collections.emptyMap();
+        }
+    }
+
+    /**
+     * Creates the keywords for searching in correction messages. This is a
+     * double-truncated search, i.e. any substring.
+     * 
+     * @param comments
+     *            the comments of a process
+     * @return keywords
+     */
+    private static final Set<String> initCommentKeywords(List<Comment> comments) {
+        Set<String> tokens = new HashSet<>();
+        for (Comment comment : comments) {
+            String message = comment.getMessage();
+            if (StringUtils.isNotBlank(message)) {
+                for (String splitWord : splitValues(message)) {
+                    String word = normalize(splitWord);
+                    for (int i = 0; i < word.length(); i++) {
+                        for (int j = i + 1; j <= word.length(); j++) {
+                            tokens.add(word.substring(i, j));
+                        }
+                    }
+                }
+            }
+        }
+        return tokens;
+    }
```

Comment:
Can you explain the reasoning here? Do i read this correctly and you are indexing every possible substing of all the words in a comment? Is this really necessary? Normal indexing should make it possible to search by word? Is this not enough?

Reply:
The [wiki page](/kitodo/kitodo-production/wiki/Suche-und-Filter) requires:

> Eine Suche nach `Meldung` findet alle Vorgänge, die einen Kommentar mit einer Nachricht wie `Das ist eine Korrekturmeldung!` haben.

To achieve this, we need both-side subword search. I only follow the expected behaviour here. I would prefer a simple-word search, too, but this is what the requirement actually looks like. Maybe we can discuss this, if it is really needed, as it inflates the index. I don’t know a clever way of splitting a German compound-word to its single words, to index just these. (This is possible, but it would be a sophisticated task of computational linguistics, typically involving a dictionary file, too, that I don’t have.)